core-primitives/stf-state-handler/src/file_io.rs

```diff
@@ -18,134 +18,364 @@
+		}
+
+		fn remove(&self, shard_identifier: &ShardIdentifier, state_id: StateId) -> Result<()> {
+			fs::remove_file(state_file_path(shard_identifier, state_id))
+				.map_err(|e| Error::Other(e.into()))
+		}
+
+		fn shard_exists(&self, shard_identifier: &ShardIdentifier) -> bool {
+			shard_exists(shard_identifier)
+		}
+
+		fn list_shards(&self) -> Result<Vec<ShardIdentifier>> {
+			list_shards()
+		}
+
+		fn list_state_ids_for_shard(
+			&self,
+			shard_identifier: &ShardIdentifier,
+		) -> Result<Vec<StateId>> {
+			let shard_path = shard_path(shard_identifier);
+			let directory_items = list_items_in_directory(&shard_path);
+
+			Ok(directory_items
+				.iter()
+				.flat_map(|item| {
+					extract_state_id_from_file_name(item.as_str())
+						// Maybe there is a better way? Need to call a function in case of `None`.
+						.ok_or_else(|| {
+							warn!("Found item ({}) that does not match state snapshot naming pattern, ignoring it", item)
+						})
+						.ok()
+				})
+				.collect())
+		}
+	}
 }
 
-pub(crate) fn exists(shard: &ShardIdentifier) -> bool {
-	Path::new(&format!("{}/{}/{}", SHARDS_PATH, shard.encode().to_base58(), ENCRYPTED_STATE_FILE))
-		.exists()
+/// Remove a shard directory with all of its content.
+pub fn remove_shard_dir(shard: &ShardIdentifier) {
+	let shard_dir_path = shard_path(shard);
+	if let Err(e) = std::fs::remove_dir_all(&shard_dir_path) {
+		error!("Failed to remove shard directory {:?}: {:?}", shard_dir_path, e);
+	}
 }
 
-pub(crate) fn init_shard(shard: &ShardIdentifier) -> Result<()> {
-	let path = format!("{}/{}", SHARDS_PATH, shard.encode().to_base58());
-	fs::create_dir_all(path.clone())?;
-	let mut file = fs::File::create(format!("{}/{}", path, ENCRYPTED_STATE_FILE))?;
-	Ok(file.write_all(b"")?)
+pub(crate) fn state_file_path(shard: &ShardIdentifier, state_id: StateId) -> PathBuf {
+	let mut shard_file_path = shard_path(shard);
+	shard_file_path.push(to_file_name(state_id));
+	shard_file_path
 }
 
-pub(crate) fn read(path: &str) -> Result<Vec<u8>> {
-	let mut bytes = io_read(path)?;
+pub(crate) fn shard_path(shard: &ShardIdentifier) -> PathBuf {
+	PathBuf::from(format!("{}/{}", SHARDS_PATH, shard.encode().to_base58()))
+}
 
-	if bytes.is_empty() {
-		return Ok(bytes)
-	}
+fn to_file_name(state_id: StateId) -> String {
+	format!("{}_{}", state_id, ENCRYPTED_STATE_FILE)
+}
 
-	let state_hash = rsgx_sha256_slice(&bytes)?;
-	debug!(
-		"read encrypted state with hash {:?} from {}",
-		H256::from_slice(state_hash.as_ref()),
-		path
-	);
+#[allow(unused)]
+fn extract_state_id_from_file_name(file_name: &str) -> Option<StateId> {
+	let state_id_str = file_name.strip_suffix(format!("_{}", ENCRYPTED_STATE_FILE).as_str())?;
+	state_id_str.parse::<StateId>().ok()
+}
 
-	AesSeal::unseal().map(|key| key.decrypt(&mut bytes))??;
-	trace!("buffer decrypted = {:?}", bytes);
+#[allow(unused)]
+fn file_for_state_exists(shard: &ShardIdentifier, state_id: StateId) -> bool {
+	state_file_path(shard, state_id).exists()
+}
```

Comment:
These functions are in use - but in the `sgx` feature only, that's why clippy reports them as 'unused'.

Reply:
I'd  prefer them feature-gated with sgx then. I can't remember precisely but at least one of cargo warnings, `cargo fmt` or `clippy` skips `#[allow(unused)]` sections.

Reply:
It is used in `cargo test`, that's why it's there, not in `sgx`. Clippy probably doesn't skip those sections, since it reports them in the first place?

Reply:
I just noticed that when I removed a huge section of `#[allow(unused)]` in another codebase, that I suddenly got tons of warnings of some sort within that 'unused code'. But how about just putting them in the test module or annotate them with `#[cfg(test)]`?

Reply:
That won't work, because they're not only test related. They're used in both the `sgx` feature, and `#[cfg(test)]` - that's what makes this tricky. (sorry the comment before might have been ambiguous)

Reply:
Haha, I see, but in that case you can do:

`#[cfg(any(test, feature = "sgx"))]`

Reply:
ah okay, good to know, I'll try that first thing tomorrow 😄